jpype/imports.py

```diff
@@ -0,0 +1,314 @@
+
+    # Is this module also a class, if so insert class info
+    jclass = _getJavaClass(javaname)
+    if jclass:
+        # Mark this as a class (will cause children to be inner classes)
+        members['__javaclass__'] = jclass
+
+        # Exposed static members as part of the module
+        _copyProperties(members, jclass.__metaclass__)
+        _copyStaticMethods(members, jclass)
+
+    return type("module." + spec.name, tuple(bases), members)
+
+def _JModule(spec, javaname):
+    """ (internal) Front end for creating a java module dynamically """
+    cls = _JImportFactory(spec, javaname)
+    out = cls(spec.name)
+    return out
+
+# %% Finder
+class _JImportLoader:
+    """ (internal) Finder hook for importlib. """
+    def find_spec(self, name, path, target):
+        parts = name.split('.', 1)
+        if not parts[0] in _JDOMAINS:
+            return None
+
+        # Support for external modules in java tree
+        for customizer in _CUSTOMIZERS:
+            if customizer.canCustomize(name):
+                return customizer.getSpec(name)
+
+        # Import the java module
+        return _ModuleSpec(name, self)
+
+    """ (internal) Loader hook for importlib. """
+    def create_module(self, spec):
+        global _initialized
+        if not _initialized:
+            raise RuntimeError("Attempt to create modules without jvm")
+
+        # Handle creating the java name based on the path
+        parts = spec.name.split('.')
+        if len(parts) == 1:
+            return _JModule(spec, _JDOMAINS[spec.name])
+
+        # Use the parent module to simplify name mangling
+        base = _sys.modules[".".join(parts[:-1])]
+
+        # Support of inner classes
+        jbasename = object.__getattribute__(base, '__javaname__')
+        try:
+            object.__getattribute(base, '__javaclass__')
+            javaname = "$".join([jbasename, _keywordUnwrap(parts[-1])])
+        except AttributeError:
+            javaname = ".".join([jbasename, _keywordUnwrap(parts[-1])])
+
+        return _JModule(spec, javaname)
+
+    def exec_module(self, fullname):
+        pass
+
+# Install hooks into python importlib
+_sys.meta_path.append(_JImportLoader())
+
+# %% Domains
+_JDOMAINS = {}
+
+def registerDomain(mod, alias=None):
+    """ Add a java domain to python as a dynamic module.
+
+    Args:
+        mod is the name of the dynamic module
+        alias is the name of the java path. (optional)
+    """
+    if not alias:
+        alias = mod
+    _JDOMAINS[mod] = alias
+
+# Preregister common top level domains
+registerDomain('com')
+registerDomain('gov')
+registerDomain('java')
+registerDomain('org')
+
+# %% Initialize
+def _initialize():
+    global _exportTypes
+    global _initialized 
+    global _modifier
+    _initialized = True
+    _JMethod = type(_JClass('java.lang.Class').forName)
+    _modifier = _JClass('java.lang.reflect.Modifier')
+    _exportTypes = (property, _JavaClass, _JImport, _JMethod)
+
+_jinit(_initialize)
```

Comment:
It is currently placed as an optional module.  To use it "import jpype; import jpype.imports".  If the support hooks are missing it would give a ImportError.  If you have an alternative mechanism I would be happy to transform it.